src/test/java/com/commercetools/sync/commons/utils/RetryableSphereClientWithExponentialBackoffTest.java

```diff
@@ -0,0 +1,278 @@
+                SphereClientConfig.of("project-key", "client-id", "client-secret");
+        final RetryableSphereClientWithExponentialBackoff retryableSphereClientWithExponentialBackoff =
+                RetryableSphereClientWithExponentialBackoff.of(clientConfig);
+        final SphereClient mockSphereUnderlyingClient =
+            spy(createHttpTestDouble(intent -> HttpResponse.of(HttpStatusCode.SERVICE_UNAVAILABLE_503)));
+
+        final long maxRetryAttempt = 2L;
+        final Function<RetryContext, Duration> durationFunction = retryContext -> Duration.ofSeconds(2);
+
+        final SphereClient decoratedSphereClient = retryableSphereClientWithExponentialBackoff.decorateSphereClient(
+                mockSphereUnderlyingClient, maxRetryAttempt, durationFunction, DEFAULT_MAX_PARALLEL_REQUESTS);
+
+        final CustomerUpdateCommand customerUpdateCommand = getCustomerUpdateCommand();
+
+        assertThat(decoratedSphereClient.execute(customerUpdateCommand))
+            .failsWithin(5, TimeUnit.SECONDS) // first retry will be in 2 second, see: durationFunction.
+            .withThrowableOfType(ExecutionException.class)
+            .withCauseExactlyInstanceOf(ServiceUnavailableException.class)
+            .withMessageContaining("503");
+
+        // first request + retries
+        verify(mockSphereUnderlyingClient, times(3)).execute(customerUpdateCommand);
+    }
+
+    @Test
+    void of_withRetryDecorator_ShouldRetryWhen504HttpResponse() {
+        final SphereClientConfig clientConfig =
+                SphereClientConfig.of("project-key", "client-id", "client-secret");
+        final RetryableSphereClientWithExponentialBackoff retryableSphereClientWithExponentialBackoff =
+                RetryableSphereClientWithExponentialBackoff.of(clientConfig);
+        final SphereClient mockSphereUnderlyingClient =
+            spy(createHttpTestDouble(intent -> HttpResponse.of(HttpStatusCode.GATEWAY_TIMEOUT_504)));
+
+        final long maxRetryAttempt = 3L;
+        final Function<RetryContext, Duration> durationFunction = retryContext -> Duration.ofSeconds(1);
+
+        final SphereClient decoratedSphereClient = retryableSphereClientWithExponentialBackoff.decorateSphereClient(
+                mockSphereUnderlyingClient, maxRetryAttempt, durationFunction, DEFAULT_MAX_PARALLEL_REQUESTS);
+
+        final CustomerUpdateCommand customerUpdateCommand = getCustomerUpdateCommand();
+
+        assertThat(decoratedSphereClient.execute(customerUpdateCommand))
+            .failsWithin(4, TimeUnit.SECONDS)
+            .withThrowableOfType(ExecutionException.class)
+            .withCauseExactlyInstanceOf(GatewayTimeoutException.class)
+            .withMessageContaining("504");
+
+        verify(mockSphereUnderlyingClient, times(4)).execute(customerUpdateCommand);
+    }
+
+    @Test
+    void of_withRetryDecorator_ShouldNotRetryWhen400HttpResponse() {
+        final SphereClientConfig clientConfig =
```

Comment:
Lines between 212-216 are the same for all tests, so you might avoid duplication. So cleaner test cases.

Reply:
Refactored. https://github.com/commercetools/commercetools-sync-java/pull/623/commits/9cb29c4ae85912b4a15a74dd1282c21f3817c5f8

Reply:
why not an instance of the test class?